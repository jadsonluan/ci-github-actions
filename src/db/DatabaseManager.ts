import { Connection, getConnection } from "typeorm";
import db from ".";
import { NODE_ENV, DB_ENV } from "../config/environment";
import Movie from "../entities/Movie";
import fs from 'fs'

export default abstract class DatabaseManager {
  connection: Connection | undefined

  constructor(conn?: Connection) {
    this.connection = conn;
  }

  async initDB() {
    this.connection = await db.init()
    return this.connection
  }

  async closeDB() {
    await this.connection.close()
  }

  getDBManager() {
    if (this.connection) {
      return this.connection.manager
    }
    return getConnection(NODE_ENV).manager
  }

  getMovie() {
    return this.getDBManager().getRepository(Movie)
  }
}

export class EntityBaseServiceManager extends DatabaseManager {
  getMigrationFileNames() {
    const list: string[] = [];
    fs.readdirSync(DB_ENV.MIGRATIONS_FOLDER, {}).forEach((file) =>
      list.push(file.toString())
    );
    return list;
  }

  private async getMigration(
    name: string
  ): Promise<{ [key: string]: any } | undefined> {
    if (/.*(.ts|.js)$/.test(name)) {
      return import(`${DB_ENV.MIGRATIONS_FOLDER}/${name}`);
    } else {
      return undefined;
    }
  }

  /**
   * Custom function for run migrations
   * Experimental
   * WIP
   */
  async runMigrations() {
    const files = this.getMigrationFileNames();
    const queryRunner = this.getDBManager().connection.createQueryRunner();
    for (const name of files) {
      const migrationsObject = await this.getMigration(name);
      if (migrationsObject) {
        const keys = Object.keys(migrationsObject);
        for (const migrationKey of keys) {
          await new migrationsObject[migrationKey]()?.up(queryRunner);
        }
      }
    }
  }

  async flushDatabase() {
    const connection = await this.getDBManager().connection;

    await connection
      .createQueryRunner()
      .dropSchema(DB_ENV.DATABASE_SCHEMA, true, true);

    await connection
      .createQueryRunner()
      .createSchema(DB_ENV.DATABASE_SCHEMA, true);

    await connection.runMigrations();
  }
}
