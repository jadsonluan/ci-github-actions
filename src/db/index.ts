import { createConnection } from "typeorm";
import { DB_ENV, NODE_ENV } from "../config/environment";
import entities from "../entities";

export default {
  init: () =>
    createConnection({
      name: NODE_ENV,
      type: 'postgres',
      synchronize: false,
      migrationsRun: false,
      host: DB_ENV.DATABASE_HOST,
      port: DB_ENV.DATABASE_PORT ? parseInt(DB_ENV.DATABASE_PORT) : 5432,
      username: DB_ENV.DATABASE_USER,
      password: DB_ENV.DATABASE_PASSWORD,
      database: DB_ENV.DATABASE_NAME,
      schema: DB_ENV.DATABASE_SCHEMA,
      entities: Object.values(entities),
      migrations: [`${DB_ENV.MIGRATIONS_FOLDER}/*.{ts,js}`],
      cli: {
        entitiesDir: DB_ENV.CLI.ENTITIES_FOLDER,
        migrationsDir: DB_ENV.CLI.MIGRATIONS_FOLDER,
      },
      logging: DB_ENV.DATABASE_LOG,
      useUTC: true,
    }),
};