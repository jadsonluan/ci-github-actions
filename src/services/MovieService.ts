import DatabaseManager from "../db/DatabaseManager";
import Movie from "../entities/Movie";


async function get(id?: number | string): Promise<Movie> {
  if (!id) {
    throw Error("id not provided");
  }

  if (typeof (id) === 'string') id = Number(id)

  return DatabaseManager.getMovie().findOne(id)
}

async function create(data: Movie): Promise<Movie> {
  if (!data.name) {
    throw Error("name not provided");
  }

  const createdEntity = DatabaseManager.getMovie().create(data)
  console.log(createdEntity)
  return DatabaseManager.getMovie().save(createdEntity)
}

async function list(limit = 10, page = 1): Promise<Movie[]> {
  if (limit < 0) limit = 10
  if (page <= 0) page = 1
  return DatabaseManager.getMovie().find({ skip: (page - 1) * limit, take: limit });
}

async function update(id, data: Movie): Promise<boolean> {
  if (!id) {
    throw Error("id not provided");
  }

  if (!data.name) return false;

  try {
    DatabaseManager.getMovie().save({ id, ...data })
    return true
  } catch (error) {
    return false
  }
}

export default {
  create,
  get,
  list,
  update
}