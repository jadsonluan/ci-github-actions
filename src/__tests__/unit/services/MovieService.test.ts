import db from "../../../db"
import MovieService from "../../../services/MovieService"

beforeEach(async () => {
  await db.init()
})

afterEach(async () => {
  await db.close()
})

describe('MovieService', () => {
  it('should show a empty list when there is no movies', async () => {
    const movies = await MovieService.list()
    expect(movies).toBeDefined()
    expect(movies).toHaveLength(0)
  })

  it('should create movies', async () => {
    const movie1 = await MovieService.create({ name: 'Shrek'})
    const movie2 = await MovieService.create({ name: 'Shrek 2'})

    expect(movie1.name).toBe('Shrek')
    expect(movie2.name).toBe('Shrek 2')
  })

  it('should not create movies without name', async () => {
    const fx = async () => await MovieService.create({ name: undefined })
    expect(fx).rejects.toThrow()
  })

  it('should list movies', async () => {
    const movies = await MovieService.list()
    expect(movies).toBeDefined()
    expect(movies).toHaveLength(2)
    expect(movies).toMatchSnapshot()
  })

  it('should list movies using pagination', async () => {
    let limit = 1
    let page = 1
    let movies = await MovieService.list(limit, page)
    expect(movies).toBeDefined()
    expect(movies).toHaveLength(1)

    page = 2
    movies = await MovieService.list(limit, page)
    expect(movies).toBeDefined()
    expect(movies).toHaveLength(1)

    page = 1
    limit = 2
    movies = await MovieService.list(limit, page)
    expect(movies).toBeDefined()
    expect(movies).toHaveLength(2)
  })

  it('should get a movie by its id', async () => {
    const movie = await MovieService.get(1)
    expect(movie.id).toBe(1)
    expect(movie.name).toBe('Shrek')
  })

  it('should not get a movie if its id is undefined', async () => {
    const fx = async () => await MovieService.get(undefined)
    expect(fx).rejects.toThrow()
  })

  it('should update a movie', async () => {
    const response = await MovieService.update(1, { name: 'Shrek 1'})
    const movie = await MovieService.get(1)
    expect(response).toBe(true)
    expect(movie.name).toBe('Shrek 1')
  })

  it('should not update movie without name', async () => {
    const fx = async () => await MovieService.update(1, { name: undefined })
    expect(fx).rejects.toThrow()
  })

  it('should not update movie without a valid id', async () => {
    const fx = async () => await MovieService.update(undefined, { name: 'Shrek 3' })
    expect(fx).rejects.toThrow()
  })
})