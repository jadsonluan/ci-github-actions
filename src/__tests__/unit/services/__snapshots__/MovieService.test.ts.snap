// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieService should list movies 1`] = `
Array [
  Movie {
    "id": 1,
    "name": "Shrek",
  },
  Movie {
    "id": 2,
    "name": "Shrek 2",
  },
]
`;
